### ISSUE TITLE
Crash due to integer multiplication overflow when initializing the backward_weights of convolution.

### ISSUE THREAD

**Issue by rfg-gh** [oneapi-src/oneDNN#1106]
# Summary
A crash happens due to an integer multiplication overflow when initializing the backward_weights of a convolution.

# Version
dnnl-1.2

# Environment
The crash happens in Linux and Windows.

# Steps to reproduce
The issue is caused by an integer multiplication overflow in jit_avx2_1x1_convolution_bwd_weights_t : init_balancers. The overflow happens after the dnnl function jit_avx2_1x1_conv_kernel_f32::init_conf is called when initializing the weights for the backward pass. Below explained how the multiplication overflow happens:

//  jit_avx2_1x1_conv_kernel_f32::init_conf generates the 2 values:
// jcp.mb and jcp.nb_reduce, that are stored and later multiplied by
//  jit_avx2_1x1_convolution_bwd_weights_t : init_balancers.
//  src_d dimensions are 268435456 x 3 x 56 x 56
//  dst_d dimensions are 268435456 x 64 x 56 x 56
//  jcp.prop_kind=backward_weights
status_t jit_avx2_1x1_conv_kernel_f32::init_conf(jit_1x1_conv_conf_t &jcp, ...)
{
    ....
    jcp.mb = src_d.dims()[0];
    jcp.oh = (ndims == 3) ? 1 : dst_d.dims()[2];
    jcp.ow = dst_d.dims()[ndims - 1];
    jcp.os = jcp.oh * jcp.ow;
    ....
    // for (jcp.prop_kind == backward_weights)
    jcp.reduce_dim = jcp.os;
    jcp.reduce_block = 1;
    ....
    jcp.nb_reduce = div_up(jcp.reduce_dim, jcp.reduce_block);
    ....
}
// Output: jcp.mb=268435456 and  jcp.nb_reduce=3136

// jit_avx2_1x1_convolution_bwd_weights_t : init_balancers produces the
// multiplication overflow: jcp_.mb * jcp_.nb_reduce, passed as argument
// to reducer_wei_conf_.init which expects this argument to be an int.
// jcp_.mb=268435456
// jcp_.nb_reduce=3136
// 268435456 * 3136 = 841813590016, Hex: [000000C4] [00000000]
// reducer_wei_conf_.init receives 0 instead, because (int)841813590016=0
// Suggestion before calling reducer_wei_conf_.init a corresponding check
// should be done for jcp_.mb*jcp_.nb_reduce to ensure that there is no
// int overflow, and this way avoid the crash.
jit_avx2_1x1_convolution_bwd_weights_t : init_balancers()
{
    ....
    reducer_wei_conf_.init(
            reduce_balancer_t(max_threads, job_size, njobs_y * njobs
                    jcp_.mb * jcp_.nb_reduce, max_buffer_size, true)
            job_size / nb_oc_blocking, nb_oc_blocking, ic_block,
            nb_ic * ic_block * oc_block, nb_oc);
    ....
}

# Observed behavior
Application crashes due to missing error handling for jcp_.mb * jcp_.nb_reduce fitting an integer.

# Expected behavior
Application should not crash, but return an error or exception.

**Comment by alokbakshi** [comment]
Hi @rfg-gh 

Running the following `benchdnn` command for latest master branch
```sh
 DNNL_MAX_CPU_ISA=AVX2 ./benchdnn  --conv --mem-check=false --dir=BWD_W mb268435456ic3oc64ih56kh1
```

(possibly owing to overflow error) shows the following assertion error
```sh
benchdnn: src/cpu/x64/cpu_reducer.cpp:40: void dnnl::impl::cpu::x64::reduce_balancer_t::balance(): Assertion `nthr_ > 0 && job_size_ > 0 && njobs_ > 0 && reduction_size_ > 0' failed.
Aborted
```

This issue will be fixed soon. Thanks for your finding!

**Comment by vpirogov** [comment]
Fixed in rls-v2.3 (c9b25e9a090d0e75710ddbfdc362fbe4c82d813e) and master (57d315ebd773a6a4e43d12a6dc47894ff8d3a326).
